gobblin-service/src/main/java/org/apache/gobblin/service/modules/flow/FlowGraphPathFinder.java

```diff
@@ -0,0 +1,428 @@
+        //       edge E'. If yes, add the edge E' to the edge queue.
+        // If the edge E' satisfies 1 and 2, add it to the edge queue for further consideration.
+        while (!edgeQueue.isEmpty()) {
+          FlowEdgeContext flowEdgeContext = edgeQueue.pop();
+
+          DataNode currentNode = this.flowGraph.getNode(flowEdgeContext.getEdge().getEndPoints().get(1));
+          DatasetDescriptor currentOutputDatasetDescriptor = flowEdgeContext.getOutputDatasetDescriptor();
+
+          //Are we done?
+          if (isPathFound(currentNode, destNode, currentOutputDatasetDescriptor, destDatasetDescriptor)) {
+            return buildDag(flowEdgeContext);
+          }
+
+          //Expand the currentNode to its adjacent edges and add them to the queue.
+          List<FlowEdgeContext> nextEdges =
+              getNextEdges(currentNode, currentOutputDatasetDescriptor, destDatasetDescriptor, flowConfig);
+          for (FlowEdgeContext childFlowEdgeContext : nextEdges) {
+            //Add a pointer from the child edge to the parent edge, if the child edge is not already in the
+            // queue.
+            if (!this.pathMap.containsKey(childFlowEdgeContext)) {
+              edgeQueue.add(childFlowEdgeContext);
+              this.pathMap.put(childFlowEdgeContext, flowEdgeContext);
+            }
+          }
+        }
+      }
+      //No path found. Return an empty dag.
+      return new Dag<>(new ArrayList<>());
+    } catch (SpecNotFoundException | JobTemplate.TemplateException | IOException | URISyntaxException e) {
+      throw new PathFinderException(
+          "Exception encountered when computing path from src: " + srcNodeId + " to dest: " + destNodeId, e);
+    }
+  }
+
+  private boolean isPathFound(DataNode currentNode, DataNode destNode, DatasetDescriptor currentDatasetDescriptor,
+      DatasetDescriptor destDatasetDescriptor) {
+    if ((currentNode.equals(destNode)) && (currentDatasetDescriptor.equals(destDatasetDescriptor))) {
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * A helper method that sorts the {@link FlowEdge}s incident on srcNode based on whether the FlowEdge has an
+   * output {@link DatasetDescriptor} that is compatible with the targetDatasetDescriptor.
+   * @param dataNode
+   * @param currentDatasetDescriptor Output {@link DatasetDescriptor} of the current edge.
+   * @param destDatasetDescriptor Target {@link DatasetDescriptor}.
+   * @return prioritized list of {@link FlowEdge}s to be added to the edge queue for expansion.
+   */
+  private List<FlowEdgeContext> getNextEdges(DataNode dataNode, DatasetDescriptor currentDatasetDescriptor,
+      DatasetDescriptor destDatasetDescriptor, Config userConfig) {
+    List<FlowEdgeContext> prioritizedEdgeList = new LinkedList<>();
+    for (FlowEdge flowEdge : this.flowGraph.getEdges(dataNode)) {
+      try {
+        DataNode edgeDestination = this.flowGraph.getNode(flowEdge.getEndPoints().get(1));
```

Comment:
`flowEdge.getEndPoints().get(1)` looks like a hacky way to represent src and dest data nodes of a flow edge. We should just define `String getSrcNodeId()` and `String getDestNodeId()`

Reply:
Agreed. Changed it to getSrc() and getDest().